src/platform/posix/posix_clock.c

```diff
@@ -19,13 +19,66 @@
 
 #ifndef NNG_USE_GETTIMEOFDAY
 
+#if __APPLE__
+#include <mach/mach.h>
+#include <mach/mach_time.h>
+#endif
+
 // Use POSIX realtime stuff
 nni_time
 nni_clock(void)
 {
 	struct timespec ts;
 	nni_time        msec;
 
+#if __APPLE__
+
```

Comment:
I really dislike `goto`s.  I guess it comes from having written some assembly in my day and seeing how easy it is to abuse... I mean, there are some cool things, like ad-hoc exception handling, but, boy is it fragile! :)

We could make a helper macro or inline function for `clock_gettime() stuff` and then use it twice, to avoid code duplication.

I'm also "allergic" to playing tricks with standard libraries symbols. There's too much of a chance of there being some macro or "intrinsic" compiler magic in there.

Reply:
Well to reduce any duplicated code the basic flow is correct.

```
    ...
#if defined(__APPLE__)
    if( <detect alt implementation needed > ) {
          // alternate implementation
    } else
#endif
  ...
```

There are two main options for how to handle the if case
a) always use the alternate if __MAC_OS_X_VERSION_MIN_REQUIRED < __MAC_10_12
b) conditionally use the alternate if the runtime check < 10.12

Reply:
I did the `static inline` common code optimisation :)

Reply:
[2 cents; still approved]

I personally prefer the
```
#if defined(__APPLE__)
    if( <detect alt implementation needed > ) {
          // alternate implementation
    } else
#endif
```
- this looks and feels like a ox12 patch, not the whole function refactoring. so it is encapsulated and can be easily removed. sure, not always this would be possible, so in general I am not criticizing your approach
- after this refactoring, we really "added a logic" that people now need to read and understand. while with the patch it is like "oh, I am not on apple, I can ignore"
- with the patch, the intention seems to be written clearer: "the main flow remains the same. but here is a special condition that will do something alternative"

Reply:
this last version looks cool to me!